### DISCUSSION TITLE
Is it possible to rename a query param segment when it has invalid rust identifier symbols?

### DISCUSSION THREAD

**Discussion by iferc** [rwf2/Rocket#2106] [Questions]
Hello!

I have no choice but needing to support query parameters that start with dollar signs `$` which means I can't access those segments in the convenient way since I end up without a valid identifier.

So far I've gotten around that with [`Request.get_query_value`](https://api.rocket.rs/v0.4/rocket/struct.Request.html#method.get_query_value) in a [`FromRequest`](https://api.rocket.rs/v0.4/rocket/request/trait.FromRequest.html), and I see that in the latest release it looks like [`Request.query_value`](https://api.rocket.rs/master/rocket/request/struct.Request.html#method.query_value) is the next variant of that.

My latest problem is that I need get these query parameters case insensitively, which isn't possible with either of those two standalone functions above. Plus I see the warning in the docs of how this is not an efficient way of pulling these query parameters which totally makes sense.

Another way I've gotten around this is with [`FromForm`](https://api.rocket.rs/v0.4/rocket/request/trait.FromForm.html) which I imagine probably has a similar effect on efficiency as the other method since I have to iterate over all query parameters to grab the ones I require. I haven't tried the latest version variant of [`FromForm`](https://api.rocket.rs/master/rocket/form/trait.FromForm.html) yet, but on the surface I'm not sure that it is a viable solution anymore.

This does solve my case insensitivity problem (in v0.4 at least) but still requires me to iterate over all query parameters.

Ideally I'd really love to have a way to basically rename a query parameter segment so that I can capture these special symbols. Whether that is in the URL string or a separate decorator like one of the following (I have no preferences on syntax):

```rs
#[get("/?<top = $top>")]
pub fn route<'a>(
    top: i64,
```

```rs
#[get("/?<$top => top>")]
pub fn route<'a>(
    top: i64,
```

```rs
#[get("/?<$top>")]
pub fn route<'a>(
    #[rocket::segment("$top")]
    top: i64,
```

If there is no way of doing the above, then it would at least be nice to have a version of [`Request.query_value`](https://api.rocket.rs/master/rocket/request/struct.Request.html#method.query_value) that could take a closure instead of a string so that I could do my own case insensitive string matching.

I think I'd be able to figure out adding a new [`Request.query_value`](https://api.rocket.rs/master/rocket/request/struct.Request.html#method.query_value) that takes a closure instead of value, but the other solution is really what I'd prefer.

- Is my solution with [`FromForm`](https://api.rocket.rs/v0.4/rocket/request/trait.FromForm.html) going to be possible in the latest version?
- Are there any ways to get around my constraint of needing invalid Rust identifiers as query segments?
- If not, could you give advice on how I might attempt to solve it myself? (I'm not sure I can but I am down to try)

Thanks!

**Comment by SergioBenitez** [comment]
> Is my solution with [FromForm](https://api.rocket.rs/v0.4/rocket/request/trait.FromForm.html) going to be possible in the latest version?

Yes. This is the preferred solution, meets all of your requirements as far as I can discern, and is efficient.
